File: java/bob/src/main/java/Bob.java
import java.text.Normalizer;
import java.util.regex.Pattern;

public class Bob {

    private static final String EMPTY_RESPONSE = "Fine. Be that way!";
    private static final String YELL_RESPONSE = "Whoa, chill out!";
    private static final String QUESTION_RESPONSE = "Sure.";
    private static final String NORMAL_RESPONSE = "Whatever.";

    public String hey(String s) {
        if (isEmpty(s)) {
            return EMPTY_RESPONSE;
        } else if (isYelling(s)) {
            return YELL_RESPONSE;
        } else if (isQuestion(s)){
            return QUESTION_RESPONSE;
        }
        return NORMAL_RESPONSE;
    }

    private boolean isEmpty(String s) {
        return s.matches("\\s*");
    }

    private boolean isYelling(String s) {
        String normalized = Normalizer.normalize(s, Normalizer.Form.NFD);
        return Pattern.compile("[A-Z]").matcher(normalized).find() && !Pattern.compile("[a-z]").matcher(normalized).find();
    }

    private boolean isQuestion(String s) {
        return s.endsWith("?");
    }
}

File: go/bob/bob.go
// Package bob provides functionality for interacting with 'bob', a lackadaisical teenager
package bob

import "strings"

// Hey responds with a response based upon a provided greeting
func Hey(remark string) string {
	switch {
	case isShouting(remark) && isQuestion(remark):
		return "Calm down, I know what I'm doing!"
	case isShouting(remark):
		return "Whoa, chill out!"
	case isQuestion(remark):
		return "Sure."
	case isSilence(remark):
		return "Fine. Be that way!"
	default:
		return "Whatever."
	}
}
func isSilence(remark string) bool {
	return strings.TrimSpace(remark) == ""
}
func isQuestion(remark string) bool {
	return strings.HasSuffix(strings.TrimSpace(remark), "?")
}
func isShouting(remark string) bool {
	// Kind of a ugly way to check the case is all caps, not just entirely composed of characters lacking a letter-case
	return remark == strings.ToUpper(remark) && remark != strings.ToLower(remark)
}
